bundle/edu.gemini.itc/src/main/java/edu/gemini/itc/gnirs/Gnirs.java

```diff
@@ -110,46 +150,38 @@ public Gnirs(GnirsParameters gp, ObservationDetails odp) {
 
-        // GNIRS is spectroscopy only
-        if (_mode instanceof Imaging) {
-            throw new RuntimeException("GNIRS does not support imaging.");
-        }
-
-
         _detector = new Detector(getDirectory() + "/", getPrefix(), "aladdin", "1K x 1K ALADDIN III InSb CCD");
         _detector.setDetectorPixels(DETECTOR_PIXELS);
 
-        _gratingOptics = new GnirsGratingOptics(
-                getDirectory() + "/" + getPrefix(), _grating,
-                _centralWavelength,
-                _detector.getDetectorPixels(),
-                1,
-                isLongCamera() ? LONG_CAMERA_SCALE_FACTOR : 1,
-                1);
-
-        if (_grating.equals(Disperser.D_10) && !isLongCamera())
-            throw new RuntimeException("The grating " + _grating + " cannot be used with the " +
-                    "0.15\" arcsec/pix (Short) camera.\n" +
-                    "  Please either change the camera or the grating.");
-
-        if (!(_filterUsed.equals("none")))
-            if ((_Filter.getStart() >= _gratingOptics.getEnd()) ||
-                    (_Filter.getEnd() <= _gratingOptics.getStart())) {
-                throw new RuntimeException("The " + _filterUsed + " filter" +
-                        " and the " + _grating +
-                        " do not overlap with the requested wavelength.\n" +
-                        " Please select a different filter, grating or wavelength.");
-            }
-
-
+        if (_mode instanceof Spectroscopy) {
+            _gratingOptics = new GnirsGratingOptics(
+                    getDirectory() + "/" + getPrefix(), _grating.get(),
+                    _centralWavelength,
+                    _detector.getDetectorPixels(),
+                    1,
+                    isLongCamera() ? LONG_CAMERA_SCALE_FACTOR : 1,
+                    1);
+
+            if (_grating.equals(Disperser.D_10) && !isLongCamera())
+                throw new RuntimeException("The grating " + _grating.get() + " cannot be used with the " +
+                        "0.15\" arcsec/pix (Short) camera.\n" +
+                        "  Please either change the camera or the grating.");
+
+            if (!_filterUsed.equals("none"))
+                if ((_Filter.getStart() >= _gratingOptics.getEnd()) ||
+                        (_Filter.getEnd() <= _gratingOptics.getStart())) {
+                    throw new RuntimeException("The " + _filterUsed + " filter" +
+                            " and the " + _grating.get() +
+                            " do not overlap with the requested wavelength.\n" +
+                            " Please select a different filter, grating or wavelength.");
+                }
+        }
         addComponent(_detector);
-
-
     }
 
     public edu.gemini.itc.base.Disperser disperser(final int order) {
         return new GnirsGratingOptics(
-                getDirectory() + "/" + getPrefix(), _grating,
+                getDirectory() + "/" + getPrefix(), _grating.get(),
```

Comment:
I should add that this is a more general problem than just in this one place; the reason being that (as @olesjasmirnova points out) for spectroscopy and imaging calculations different preconditions are true (e.g. disperser *must* be present for spectroscopy).

These are sometimes checked (using `Exception`s to signal the problem to the user) and sometimes not. Originally all of these optional objects (filters, disperser) were handled as present or `null`, and to make this more explicit I started to use `Option`s but never took the time to really clean this up through all levels of the call hierarchy, that's why there are some places with `get`s.

Not trying to say this is good, just trying to point out the history. I was always planning to clean this up and introduce a clearer separation between spectroscopy and imaging cases which would allow to do the checking of preconditions (disperser present? ..etc) more explicit and turn the optional objects into the actual ones in the same step and then pass those into the calculations. But I guess someone else will have to follow up with that. So please be nice with @olesjasmirnova when she uses `get` in some places ;)